src/error.rs

```diff
@@ -693,6 +698,61 @@ impl WriteFailure {
     }
 }
 
+/// An error that occurred during a GridFS operation.
+#[derive(Clone, Debug)]
+#[allow(missing_docs)]
+#[non_exhaustive]
+pub enum GridFsErrorKind {
+    /// The file with the given identifier was not found.
+    #[non_exhaustive]
+    FileNotFound { identifier: GridFsFileIdentifier },
+
+    /// The file with the given revision was not found.
+    #[non_exhaustive]
+    RevisionNotFound { revision: i32 },
+
+    /// The chunk at index 'n' was missing.
+    #[non_exhaustive]
+    MissingChunk { n: u32 },
+
+    /// The chunk was the incorrect size.
+    #[non_exhaustive]
+    WrongSizeChunk {
+        actual_size: u32,
+        expected_size: u32,
+    },
+
+    /// An incorrect number of chunks was present for the file.
+    #[non_exhaustive]
+    WrongNumberOfChunks {
+        actual_number: u32,
+        expected_number: u32,
+    },
+}
+
+/// An identifier for a file stored in a GridFS bucket.
+#[derive(Clone, Debug)]
+#[non_exhaustive]
+pub enum GridFsFileIdentifier {
+    /// The name of the file. Not guaranteed to be unique.
+    Filename(String),
+
+    /// The file's unique [`Bson`] ID.
+    Id(Bson),
+}
+
+impl From<&str> for GridFsFileIdentifier {
+    fn from(filename: &str) -> Self {
+        Self::Filename(filename.into())
+    }
+}
+
+impl From<Bson> for GridFsFileIdentifier {
+    fn from(id: Bson) -> Self {
+        Self::Id(id)
+    }
+}
```

Comment:
I think it might actually be better to omit these since a string can be used to represent either a filename or an ID. Having to explicitly indicate which variant helps avoid this ambiguity.